### DISCUSSION TITLE
Dynamic / Mutable Contexts

### DISCUSSION THREAD

**Discussion by Tkael** [r3c/cottle#237] [General]
Hi there. First off, thank you for Cottle! It's been a core library for the application I develop (https://github.com/EDCD/EDDI) for many years.

We'd like to upgrade to the latest version of the library but we stopped taking updates after version 1.4.0.4 because version 1.4.0.4 was the last release to support mutable contexts. Mutable contexts allow us to set up context variables in an outer template and then render an inner template using context variables set in the outer scope. This is a desirable feature that we do not wish to lose.

I'm taking another look at this and would like to know if you have any advice for moving to the latest version of the library?
Perhaps this may be accomplished by creating a custom parser (similar to `set`) rather than trying to update the context via a custom function?

**Comment by r3c** [comment]
Hi @Tkael and thank you for your feedback!

Indeed the mutability of contexts was removed to simplify the API. Here are a few suggestions about how you could set variables in a template and use them in another one:

# Return and inject

The most straightforward way I see is to return variables from your "source" template so you can inject in the other one. It would look a bit like this:

File `source.cottle`:
```
// Your code generating variables...

{return ["var1": var1, "var2": var2, ...]}
```

File `renderer.cs`:
```csharp
var document1 = Document.CreateDefault(/* ... */);
var context1 = Context.CreateBuiltin(/* ... */);
var writer1 = new StringWriter();
var result1 = document1.Render(context1, writer1);

// Inject variables from source document into second one:
var document2 = Document.CreateDefault(/* ... */);
var context2 = Context.CreateBuiltin(result1.Fields.ToDictionary(field => field.Key, field => field.Value));
var writer2 = new StringWriter();
var result2 = document2.Render(context2, writer2); // Or just `var output = document2.Render(context2);` as you're not interested in returned value here
```

# Custom storage

Your idea about custom function would work fine as well. You could have a pair of `setGlobal` and `getGlobal` functions acting as setter and getter to some custom storage of yours:

File `renderer.cs`:
```csharp
var storage = new Dictionary<Value, Value>();
var getGlobal = Function.CreatePure1((state, name) => storage.TryGetValue(name, out var value) ? value : Value.Undefined);
var setGlobal = Function.CreatePure2((state, name, value) =>
{
    storage[name] = value;

    return value;
});

var context = Context.CreateBuiltin(new Dictionary<Value, Value>
{
    ["getGlobal"] = getGlobal,
    ["setGlobal"] = setGlobal,
    /* ... */
});

var document1 = Document.CreateDefault(/* ... */);
var output1 = document1.Render(context);

// Pass the same `getGlobal` and `setGlobal` functions to 2nd document so it can access shared storage
var document2 = Document.CreateDefault(/* ... */);
var output2 = document2.Render(context);
```

# Include with context

That's a variant of first solution and could be acceptable depending on how your templates are accessed:

File `renderer.cs`:
```csharp
var context = Context.Empty; // For use in closure
var include = Function.Create2((state, path, variables, writer) =>
{
    var includedDocument = Document.CreateDefault(File.ReadAllText(path.AsString)); // If your templates are stored as files
    var includedContext = Context.CreateCascade(
        Context.CreateDefault(variable.Fields.ToDictionary(field => field.Key, field => field.Value)),
        context);

    return includedDocument.Render(includedContext, writer);
});

var document = Document.CreateDefault(new Dictionary<Value, Value> { ["include"] = include, /* ... */ });
var output = document.Render(context);
```

File `source.cottle`:
```
// Your code generating variables...

{include("other_template.cottle", ["var1": var1, "var2": var2, ...]}
```

I'm not sure about the option of modifying the parser, as it would require more code change and may not be done without breaking the API. Please let me know if any of these option sound acceptable to you, or otherwise feel free to share more context so I can suggest alternatives 🙂

**Comment by Tkael** [comment]
Thank you Remi, I will think on the options you've presented. The 2nd and 3rd options are probably closest to what we're currently using.

As a follow-up question, I understand that a "pure" function is a function that does not have "side effects" but I'm not sure what "side effects" would mean in the context of my custom functions. Would it be possible to provide a definition for this?

**Reply by r3c** [reply]
The documentation is indeed quite opaque about how pure functions differ from non-pure ones, I believe the only mention of it is here: https://cottle.readthedocs.io/en/latest/page/06-api.html#property-Cottle.IFunction.IsPure

What it means is that function doesn't modify any internal state nor writes to template output stream. As a consequence it can be freely reordered by the compiler without changing the result of template evaluation, and this makes it eligible for some optimizations. For example, expressions known to be always true or always false at compile time can easily be optimized away, but that only holds if everything within these expressions is pure. If your expression looks like `{if myFunction(x):y}` then even if `myFunction` is always returning `true` for known values of `x` then it can't be removed, as it could still be printing some contents to output as a consequence of being invoked.

I hope this makes things more clear, if yes I'll try improving the documentation as well 🙂

**Reply by Tkael** [reply]
I'm afraid that explanation is still a little difficult for me to follow.

So going back to your first response I think I understand that the `setGlobal` function would be pure but the `getGlobal` function might be impure (since `getGlobal` may impact the result of the template evaluation and consequently cannot be freely reordered). Is that right? 

Or would it be more correct to say that a pure function is one where the nested contents do not require further rendering while an impure function may need to pass through multiple rendering levels to render the contents and then function result fully?

**Reply by r3c** [reply]
I realize I was wrong in my previous comment, saying that `getGlobal` could be pure: a pure function doesn't access anything else than constants and the arguments it received, so `getGlobal` is impure by accessing the non-constant storage it shares with `setGlobal`. You'll be more safe declaring both of them as impure, sorry for the confusion 🙂

**Reply by Tkael** [reply]
This is a clearer definition for me, thank you.
`a pure function doesn't access anything else than constants and the arguments it received`

**Comment by Tkael** [comment]
Hi Remi,

I'm back with a specific function that I'm trying to update and debug. Please review and let me know what you think?

File `source.cottle`:
```
{set x to "Hello"} 
{set y to "Hi"}

{_ Should render as Hello world or Hi there }
{OneOf("{x} world", "{y} there")}

{_ Should render as Hello world or Hi there }
{set mapExample to ["{x} world", "{y} there"]}
{OneOf(mapexample)}

{_ Here's a rough Cottle equivalent of what this function is intended to do, albeit for a fixed number of variables }
{ set x to "Hello" }
{ set y to "Hi" }
{ set mapVars to [call(cat, [x, " world"]), call(cat, [y, " there"])]}
{ mapVars[rand(len(mapVars))] }

{_ I'd also like to be able to nest function results, e.g. the function below should render as a, b, or c.  }
{OneOf("a", OneOf('b', 'c'))}
```

File `renderer.cs`
```cs
        private static readonly Random random =
            new Random( new { n = nameof(OneOf), dt = DateTime.UtcNow }.GetHashCode() );

        var dict = new Dictionary<Value, Value> { ["OneOf"] = function , /* ... */ };
        var context = Context.CreateBuiltin( dict )

        public IFunction function => Function.Create( ( state, values, writer) =>
        {
            Value result;
            if ( values.Count == 1 && values[ 0 ].Type == ValueContent.Map )
            {
                values[ 0 ].Fields.TryGet( random.Next( values[ 0 ].Fields.Count ), out result );
            }
            else
            {
                var rand = random.Next( values.Count );
                result = values[ rand ];
            }
			var includedDocument = Document.CreateDefault(result);
            var includedContext = Context.CreateCascade(
                  Context.CreateDefault(variable.Fields.ToDictionary(field => field.Key, field => field.Value)), context);
            return includedDocument.Render(includedContext);
        } );

    var document = Document.CreateDefault(File.ReadAllText(path.AsString));
    var output = document.Render(context);
```

As you can see, at the moment I'm passing the function result to a new document. On the old version of Cottle I have been using, I was able to pass a mutated context to the includedDocument. But since the context is now immutable the values of `x` and `y` from the example above are not available in the includedDocument. Please advise on how I can approach this function (or if it is even possible for a custom function to do this in the current version)? Perhaps using the `writer`?

Thank you!

**Reply by r3c** [reply]
If I understand correctly you would like strings `"{x} world"` and `"{y} there"` to be rendered using current context and stored into variables you can use in your template? If that's correct, would something like this match your need?

```
{set a to:{x} world}
{set b to:{y} there}
{set mapExample to [a, b]}
{OneOf(mapexample)}
```

**Reply by Tkael** [reply]
If you are proposing that pattern with the current `OneOf` function I shared above then I'm afraid it It would be a major breaking change for the API we've been supporting for as long as we've been using Cottle. It would make the templates harder to read, require us to rewrite the vast majority of our templates, and would require our users, who can write their own custom templates, to do the same. :-(

Here's a more complete example of a template to show how we're using custom functions to create dynamic speech.
The `F` function renders another included document template using the template name.
The `Occasionally` function takes 2 arguments with the first argument indicating how often the second argument is rendered.
The `Pause` function adds a pause.
All of this together is processed through a Text-To-Speech processor to create dynamic spoken responses to a player's in-game actions.
```
{_ Context }
{SetState('eddi_context_last_subject', 'landing pad')}
{SetState('eddi_context_landing_pad_system', system.name)}
{SetState('eddi_context_landing_pad_station', event.station)}
{SetState('eddi_context_landing_pad_station_model', event.stationtype)}
{SetState('eddi_context_landing_pad_pad', event.landingpad)}

{for compartment in ship.compartments:
    {if find(compartment.module.name, "Docking") > -1:
        {if compartment.module.enabled:
            {set hasDockingComputerReady to true}
        }
    }
}

{if hasDockingComputerReady:
    {Pause(2000)}
    {OneOf(
        "Ready to engage {OneOf('auto dock sequence', 'automated docking')}", 
        "{OneOf('Auto dock', 'Automated docking')} {OneOf('ready', 'standing by')}"
    )}
    {Occasionally(4, cat(', ', F('Honorific') ))}.
|else:
    {if event.stationDefinition.basename = "Coriolis" || 
        event.stationDefinition.basename = "Orbis" || 
        event.stationDefinition.basename = "Bernal" || 
        event.stationDefinition.basename = "Ocellus" || 
        event.stationDefinition.basename = "AsteroidBase":
        {Pause(8000)}
        {F("Landing pad report")}
    |else:
        {Pause(8000)}
        Landing pad {event.landingpad}
        {OneOf('ready', 'assigned', 'allocated')}
        {Occasionally(4, cat(', ', F('Honorific') ))}.
    }
}
```

**Reply by Tkael** [reply]
I don't know if this could work but what if I could designate custom functions which could run ___after___ the parser resolves / executes and before the document render is finalized? Would that allow them to integrate more nicely with native functions like `set`?

EDIT: Could lazy value evaluation help?

**Comment by r3c** [comment]
I'm having a hard time understanding your previous example, especially the parts that contain nested template-in-template fragments (`{OneOf("Ready to engage {OneOf('auto dock sequence', 'automated docking')}"...`). Maybe if you give me a minimal reproduction of what you're willing to achieve I could suggest another solution?

**Reply by Tkael** [reply]
```
    {OneOf(
        "Ready to engage {OneOf('auto dock sequence', 'automated docking')}", 
        "{OneOf('Auto dock', 'Automated docking')} {OneOf('ready', 'standing by')}"
    )}
```
would randomly render as one of:
- "Ready to engage auto dock sequence"
- "Ready to engage automated docking"
- "Auto dock ready"
- "Auto dock standing by"
- "Automated docking ready"
- "Automated docking standing by"

**Reply by r3c** [reply]
I guess this wouldn't be an option?

```
{OneOf(
    cat("Ready to engage ", OneOf("auto dock sequence", "automated docking")),
    cat(OneOf("Auto dock", "Automated docking"), " ", OneOf("ready", "standing by"))
}
```

I'm not able to establish a link between this syntax and the original post about mutable contexts, so I guess I miss something.

**Reply by Tkael** [reply]
I apologize if I have not been sufficiently clear. I am trying to explain our current and historical template syntax because I do have an API that I'm trying to maintain with the people who use my application and want to minimize changes to that syntax. 

With the current template syntax we are able to `set` a variable in the outer template and then pass that state to be rendered using an `includedDocument` through the custom function. But with the upgraded version of Cottle the state of the outer template is no longer mutable and consequently we are unable to render the `includedDocument` using the mutated state.

I'm looking for a solution that removes the need for my function to render an `includedDocument` which may depend on a context and variables which are `set` in the outer document. Your proposed syntax change may allow this and if so then the change in the template syntax could still be worth exploring even if it would require a very major change. I will test this tomorrow. 

Thank you again for the feedback you have given me - it has been very helpful.

**Reply by r3c** [reply]
No worries, it's just that I was having a hard time understanding your use case exactly 🙂

Let me try to reword so I can check if I got it right this time: you implemented some function (the `OneOf` from previous example) that receives strings as arguments. This function is expected to consider the strings as Cottle templates and render them while having access to every variable available from parent document's rendering context.

Assuming this is correct, forwarding original context is fairly simple and I understand your issue is about variables being set from inside the template. Unfortunately there is no way currently to intercept what `{set x to y}`: storage of these variables is private to each document and the API doesn't offer a way to access them (mostly because some information is lost after compiling a template, for example variable names are not preserved).

The solution that seems most suitable right now is the custom setter/getter functions that would allow you to keep track of variables. If that's not a viable option for you, let me know and I'll try to think of other alternatives but for now I don't see anything that would work out of the box without any change to template files, sorry.

[edit] One idea that could maybe work, but would require a breaking change in Cottle API, is to provide access to all global variables when implementing a custom function. For example the opaque "state" variable being passed to callback in `Function.Create0((state, writer) => { /* some code */ return Value.Undefined; })` could become less opaque and be an interface allowing access to global variables. This would however change the callback type of everything passed to `Function.Create*()` so I'm not sure it's a good idea yet and need to spend a bit more of thinking about it 🙂

**Reply by Tkael** [reply]
Yes, that is a good summary of the problem. :-)

And yes, `SetState` could be used in place of `set` to achieve the effect but that would create a barrier between the native `set` function and my custom function that I'd prefer to avoid. And it would create a similar barrier between my function and any other native function which interacts with global variables.

As for the `state` object, being able to access that object and read the global state would be a great help to simplify my code to render any `includedDocument` that I might wish to render. As is, it is a little tantalizing and frustrating to pass the object through my function without being able to meaningfully interact or read from it. :-)

**Reply by r3c** [reply]
Hello,

I opened PR #239 with a proposal of API change. Non-pure functions can now be declared using one of `Function.CreateNative*` helpers. These helpers expect a callback whose first argument is a `IRuntime` instance that allows accessing global variables. Here is an example using `Function.CreateNative0` (aka a non-pure function taking 0 arguments):

render.cs:
```csharp
var context = Context.CreateCustom(new Dictionary<Value, Value>
{
    ["canAccessGlobals"] = Value.FromFunction(Function.CreateNative0((runtime, output) =>
    {
        DoSomethingWith(runtime.Globals);

        return Value.Undefined;
    }
});

document.Render(context);
```

template.cottle:
```
{set a to 1}
{set b to 2}
{canAccessGlobals()}
```

With this code, `canAccessGlobal` can access `a`, `b` and `canAccessGlobals` from within the callback definition. I'm attaching a compiled NuGet of this experimental version, but you can also compile code from the PR directly of course.

Can you please let me know if it could help addressing your requirements?

[Cottle.2.1.0-experimental1.zip](https://github.com/user-attachments/files/15936669/Cottle.2.1.0-experimental1.zip)

**Reply by Tkael** [reply]
Thank you! I will begin examining this immediately. :-)

**Reply by Tkael** [reply]
Hi there. Reporting back to say I think (pending additional testing) that all functions now *seem* to be behaving as expected. I will of course need to have my users test this more thoroughly but your experimental build has been extremely helpful. :-) 🎆 

As a secondary note, I've found that when working with an object being referenced by its base Type, Value.FromReflection seems to currently pick up the properties of the base Type only and not the properties of the underlying Type. ReflectionValue() used to pick up both. I was able to solve this by passing the object as a dynamic type to Value.FromReflection.

**Reply by r3c** [reply]
OK glad to hear it, I'll wait for your feedback about whether this change should be merged or not 🙂

About `Value.FromReflection` (maybe worth a separate issue as it sounds unrelated) I'd appreciate an example to understand what the difference is (`Value.FromReflection<T>(t)` is supposed to infer fields by dynamically scanning type `T`).

**Reply by Tkael** [reply]
Value.FromReflection seems like it may also have a problem with List\<T\>... I was getting zero results until I cast to dynamic.

**Reply by r3c** [reply]
Would you please have an example to share? I just sent #240 to make sure it works as expected and it seems `List<>` works as expected?

**Reply by Tkael** [reply]
I'm afraid I am not very familiar with the test framework that you are using but I'll look for a way to express what I was seeing.
It was a List<T> where T was an object with various properties and fields which I tried to parse with Value.FromReflection.
Cottle was behaving like that list had zero objects until I cast from List<T> to dynamic.

**Reply by Tkael** [reply]
I ran into another List<T> which `Value.FromReflection` failed to parse. I simplified the object T which resolved the parsing error (leading me to conclude that it the issue has more to do with the object T than the inclusion of the object in a List. Testing further, I found that the Type below is an example which throws an `ArgumentException` with the message `The type 'System.Void*' may not be used as a type argument.`.

DemoForm.cs
```cs
        internal class TestObject
        {
            public string test = "Test";
            public HashSet<XmlSchemaSet> SchemaSets = new() { new XmlSchemaSet() };
        }

        private void buttonEvaluate_Click(object sender, EventArgs e)
        {
            var result = Document.CreateDefault(textBoxTemplate.Text, _configuration);

            if (!result.Success)
            {
                DisplayResult(result);

                return;
            }

            var symbols = new Dictionary<Value, Value>();
            symbols["test"] = Value.FromReflection(new TestObject(),
                BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);

            foreach (TreeNode root in treeViewContext.Nodes)
            {
                foreach (var pair in DemoForm.ValuesBuild(root.Nodes))
                    symbols[pair.Key] = pair.Value;
            }

            DisplayOutput(result.Document.Render(Context.CreateBuiltin(symbols)));
        }
```

source.cottle:
```
{test.test}
```

**Reply by r3c** [reply]
OK now I understand the confusion, when you first mentioned a `List` I thought it was type `System.List<T>`, but your example seems to exhibit an issue with a different type. I'll have a look, but going back to your original question can you please tell me what I should do with the pending change on functions accessing globals?

Thank you.

[edit] I had a look at the issue related to `XmlSchemaSet`. It contains some private system fields that are too complex for being converted, namely `XmlSchemaSet._internalEventHandler` which based on the name isn't something designed to be used in client code I believe. Cottle tries to iterate every one of them because of the `BindingFlags.NonPublic` flag you passed, so you can simply work around this error by specifying `BindingFlags.Instance | BindingFlags.Public` instead.

**Reply by Tkael** [reply]
> OK now I understand the confusion, when you first mentioned a `List` I thought it was type `System.List<T>`, but your example seems to exhibit an issue with a different type. I'll have a look, but going back to your original question can you please tell me what I should do with the pending change on functions accessing globals?
> 
> Thank you.

I have tested the pending change about as thoroughly as I am capable of doing on my own. It seems quite solid. :-)
I have released an alpha and should have feedback from my testers within the next couple of days.

**Reply by r3c** [reply]
Perfect 👍 Just in case you didn't see, I updated my previous message with details about your other issue on `XmlSchemaSet`. That's something the library can probably work around as well, but in this specific case I feel accessing non-public fields may not be a safe behavior.

**Reply by Tkael** [reply]
Hi Remi, my testers have had several days now with the latest experimental build and we haven't run into any major unexpected issues. I'd like to recommend merging your experimental branch. Thank you again!

**Reply by r3c** [reply]
I just published v2.1.0 with this change, as well as an improvement for `Value.FromReflection` to avoid running into the issue you experienced when wrapping `XmlSchemaSet`. Hope that answers your needs!

https://github.com/r3c/cottle/releases/tag/2.1.0

**Reply by r3c** [reply]
Hey @Tkael, I didn't hear back from you following this discussion. Hope it eventually unblocked you and allowed you to bump the library. I'm closing this now, feel free to open a new issue of you need to follow up.

**Reply by Tkael** [reply]
Hi there Remi, I was able to bump the library. Thank you again!
